components/style/traversal.rs

```diff
@@ -334,38 +356,52 @@ pub fn recalc_style_at<'a, E, C, D>(context: &'a C,
 
                 // Perform the CSS cascade.
                 unsafe {
-                    restyle_result = element.cascade_node(context,
-                                                          element.parent_element(),
-                                                          &applicable_declarations);
+                    element.cascade_node(context, data, element.parent_element(),
+                                         &applicable_declarations);
                 }
 
                 // Add ourselves to the LRU cache.
                 if let Some(element) = shareable_element {
-                    style_sharing_candidate_cache.insert_if_possible(&element, relations);
+                    style_sharing_candidate_cache.insert_if_possible(&element,
+                                                                     &element.borrow_data()
+                                                                             .unwrap()
+                                                                             .current_styles()
+                                                                             .primary,
+                                                                     relations);
                 }
             }
-            StyleSharingResult::StyleWasShared(index, damage, cached_restyle_result) => {
-                restyle_result = cached_restyle_result;
+            StyleSharingResult::StyleWasShared(index, damage) => {
                 if opts::get().style_sharing_stats {
                     STYLE_SHARING_CACHE_HITS.fetch_add(1, Ordering::Relaxed);
                 }
                 style_sharing_candidate_cache.touch(index);
+
+                // Drop the mutable borrow early, since Servo's set_restyle_damage also borrows.
+                mem::drop(data);
+
                 element.set_restyle_damage(damage);
             }
         }
     }
 
-    // If we restyled this node, conservatively mark all our children as needing
-    // processing. The eventual algorithm we're designing does this in a more granular
-    // fashion.
-    if mode == StylingMode::Restyle && restyle_result == RestyleResult::Continue {
+    if element.is_display_none() {
+        // If this element is display:none, throw away all style data in the subtree.
+        fn clear_descendant_data<E: TElement, D: DomTraversalContext<E::ConcreteNode>>(el: E) {
```

Comment:
This seems rather unfortunate. This effectively means that if I do `document.body.style.display = "none"`, we end up traversing all the tree anyway (in a sequential fashion) dropping restyle data, right?

Is this the plan long-term, or only a work in progress? Is there anything we can do to avoid this?